### PR TITLE
Support never missing a block using the `CurrentBlockStream`

### PR DESCRIPTION
**Pull request by MartinquaXD**
# Description
The `CurrentBlockStream` only yields the current block whenever it gets polled. This is problematic if whatever task that runs based on that stream takes longer than the average block time. In that case we would completely miss 1 block.
For the vast majority of use cases this is fine but the event indexing logic also runs based on that stream and there it's extremely important that we don't miss any blocks.

# Changes
Replaced the type alias `CurrentBlockStream` with a new type. The new type is able to support both use cases (watch latest block and watch all blocks) and has an API that is more explicit in what it does.

I used the new feature that yields all the blocks for the event indexing and left the current behavior for everything since we don't need to see all blocks there.

## How to test
Should be covered by existing tests

### CODE REVIEW

**Review by MartinquaXD — COMMENTED** [review]
One thing that I just noticed is that this only helps when the event indexing logic is only sometimes slow. In that case the block buffer will fill up a bit and once the indexing logic overcomes the temporary slow down it can quickly go through the buffered blocks.
However, to never miss any blocks even if the logic is always slow we probably have to spawn a new indexing task on every new block. That way the indexing logic could theoretically take 5 minutes and we still wouldn't miss any blocks.

**Comment by MartinquaXD** [comment]
Looks like this was a false alert since the indexing logic was written such that it always checks the blocks missed since the last time the update logic ran. But with the nicer API introduced in this PR I would still consider this PR worth a look.

File: crates/ethrpc/src/current_block/mod.rs [hunk 1]
```diff
@@ -16,6 +16,130 @@ use {
...
+    /// Returns a stream that always yields the most recent block (if unseen).
+    /// Also yields the current block right away and not just the next new
+    /// block. Use this if you only need to know about the current state.
+    pub fn watch_stream(&self) -> WatchStream<BlockInfo> {
+        WatchStream::new(self.0.watch_receiver.clone())
+    }
+
+    /// Returns a stream that yields every new block **AFTER** it got created.
+    /// Use this if it's required to not miss any blocks.
+    pub fn buffering_stream(&self) -> impl Stream<Item = BlockInfo> {
```
Comment (fleupold) [inline crates/ethrpc/src/current_block/mod.rs:70]: This is bound by `MAX_SIZE` right? Let's add that as a comment.

Also we should comment on how the buffer deals with reorgs (it seems it doesn't)

File: crates/ethrpc/src/current_block/mod.rs [hunk 2]
```diff
@@ -16,6 +16,130 @@ use {
...
+                }
+
+                // The new block is different but might still have the same number.
+                tracing::debug!(number=%block.number, hash=?block.hash, "polled block");
+                update_block_metrics(previous_block.number, block.number);
+
+                // Only update the stream if the number has increased.
+                if block.number <= previous_block.number {
+                    continue;
+                }
```
Comment (fleupold) [inline crates/ethrpc/src/current_block/mod.rs:116]: I see this is where #2740 comes into play? I think we can add logic here where if the number is > previous_block.number + 1, we specifically poll the blocks from previous_block.number +1 until this one to make the stream gap free (should be a separate PR/task though as it seems somewhat involved if we also want to handle reorgs).

File: crates/ethrpc/src/current_block/mod.rs (hunk 1 above)
Comment (fleupold) [reply crates/ethrpc/src/current_block/mod.rs:70]: Also, where is this used now?

**Review by fleupold — APPROVED** [review]
Great PR!

File: crates/ethrpc/src/current_block/mod.rs (hunk 1 above)
Comment (MartinquaXD) [reply crates/ethrpc/src/current_block/mod.rs:70]: After discovering that the indexing logic actually does not need to see every single block to work correctly as I initially thought this function did not end up being used. So technically this PR could get simplified a bit again although I don't really see harm in keeping it now that we have it.

File: crates/ethrpc/src/current_block/mod.rs (hunk 2 above)
Comment (MartinquaXD) [reply crates/ethrpc/src/current_block/mod.rs:116]: Oh, right. Simply looping over all the blocks we know we missed would be way simpler than introducing web socket support. 👍

File: crates/ethrpc/src/current_block/mod.rs [hunk 3]
```diff
@@ -16,6 +16,136 @@ use {
...
+        let first_block = web3.current_block().await?;
+        tracing::debug!(number=%first_block.number, hash=?first_block.hash, "polled block");
+
+        let stream = CurrentBlockStream(Arc::new(BlockStreamInner::new(first_block)));
+        let stream_clone = stream.clone();
+
+        let update_future = async move {
+            let mut previous_block = first_block;
+            loop {
+                tokio::time::sleep(poll_interval).await;
```
Comment (squadgazzz) [inline crates/ethrpc/src/current_block/mod.rs:101]: I probably missed or misunderstood something, but if more than 1 block was mined during that period, we might still miss the intermediate ones, doesn't it?

File: crates/ethrpc/src/current_block/mod.rs [hunk 4]
```diff
@@ -16,6 +16,136 @@ use {
...
+
+/// Monitors the blockchain for new blocks and allows users to get notified
+/// about them. Can be considered the "heartbeat" of the backend.
+#[derive(Debug, Clone)]
+pub struct CurrentBlockStream(Arc<BlockStreamInner>);
+
+impl CurrentBlockStream {
+    /// Returns a reference to the current block. Keeping this reference
+    /// alive locks a mutex so it should be used as shortly as possible.
+    pub fn current(&self) -> watch::Ref<'_, BlockInfo> {
```
Comment (sunce86) [inline crates/ethrpc/src/current_block/mod.rs:57]: Now that we wrapped `.borrow()` should we consider not returning a reference here but the copy of data, so that no accidental lock is kept at callsite?

**Review by sunce86 — APPROVED** [review]
lg

File: crates/ethrpc/src/current_block/mod.rs (hunk 4 above)
Comment (MartinquaXD) [reply crates/ethrpc/src/current_block/mod.rs:57]: Fine with me. 👌

File: crates/ethrpc/src/current_block/mod.rs (hunk 3 above)
Comment (MartinquaXD) [reply crates/ethrpc/src/current_block/mod.rs:101]: Yeah, this is correct. I guess the PR description is slightly misleading. With this PR we'll add support for creating a stream that yields all blocks the stream ever saw. If the stream missed a block in the first place the subsequent streams will also not see this block. Felix suggested a good way to address this in a follow up PR though.

**Review by squadgazzz — APPROVED** [review]

File: crates/ethrpc/src/current_block/mod.rs [hunk 5]
```diff
@@ -16,6 +16,135 @@ use {
...
+                if previous_block.hash == block.hash {
+                    continue;
+                }
+
+                // The new block is different but might still have the same number.
+                tracing::debug!(number=%block.number, hash=?block.hash, "polled block");
+                update_block_metrics(previous_block.number, block.number);
+
+                // Only update the stream if the number has increased.
+                if block.number <= previous_block.number {
```
Comment (squadgazzz) [inline crates/ethrpc/src/current_block/mod.rs:119]: All the current readers don't care about reorgs, do they?

Comment (MartinquaXD) [reply crates/ethrpc/src/current_block/mod.rs:119]: Current readers only care about the latest block number.
But I think it would be good if we would detect 1 block reorgs here. Right now they would get ignored (which is what you are referring to, right?).
I would prefer to leave the behavioral change (notify about 1 block reorgs) to the next PR where I backfill gaps when we miss blocks.

Comment (sunce86) [reply crates/ethrpc/src/current_block/mod.rs:119]: Yes, we should split behavioral changes from refactoring to isolate potential bugs.
